File: tools/image2cs.js
/*
 * Copyright (c) 2016-2017  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK Tools.
 * 
 *   The Moddable SDK Tools is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 * 
 *   The Moddable SDK Tools is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 * 
 *   You should have received a copy of the GNU General Public License
 *   along with the Moddable SDK Tools.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import { FILE, TOOL } from "tool";
import Bitmap from "commodetto/Bitmap";
import ColorCellOut from "commodetto/ColorCellOut";
import Convert from "commodetto/Convert";
import JPEG from "commodetto/ReadJPEG";
import PNG from "commodetto/ReadPNG";

export default class extends TOOL {
	constructor(argv) {
		super(argv);
		this.directory = false;
		this.format = Bitmap.RGB565LE;
		this.quality = undefined;
		this.rotation = 0;
		this.inputPath = null;
		this.outputPath = null;
		this.windows = this.currentPlatform == "win";
		this.slash = this.windows ? "\\" : "/";
		var argc = argv.length;
		for (var argi = 1; argi < argc; argi++) {
			var option = argv[argi], name, path;
			switch (option) {
			case "-o":
				argi++;	
				if (argi >= argc)
					throw new Error("-o: no directory!");
				name = argv[argi];
				if (this.outputPath)
					throw new Error("-o '" + name + "': too many directories!");
				path = this.resolveDirectoryPath(name);
				if (!path)
					throw new Error("-o '" + name + "': directory not found!");
				this.outputPath = path;
				break;
			case "-q":
				argi++;	
				if (argi >= argc)
					throw new Error("-q: no quality!");
				name = parseInt(argv[argi]);
				if ((name < 0) || (100 < name))
					throw new Error("-q: " + name + ": invalid quality!");
				this.quality = name;
				break;
			case "-r":
				argi++;	
				if (argi >= argc)
					throw new Error("-r: no rotation!");
				name = parseInt(argv[argi]);
				if ((name != 0) && (name != 90) && (name != 180) && (name != 270))
					throw new Error("-r: " + name + ": invalid rotation!");
				this.rotation = name;
				break;
			default:
				name = argv[argi];
				if (this.inputPath)
					throw new Error("'" + name + "': too many files!");
				path = this.resolveFilePath(name);
				if (!path) {
					path = this.resolveDirectoryPath(name);
					if (!path)
						throw new Error("'" + name + "': file not found!");
					this.directory = true;
				}
				this.inputPath = path;
				break;
			}
		}
		if (!this.inputPath) {
			throw new Error("no file!");
		}
		if (!this.outputPath)
			this.outputPath = this.splitPath(this.inputPath).directory;
	}
	compress(buffer, width, height) {
		let writer = new ColorCellOut;
		if (this.quality !== undefined)
			writer.begin(0, 0, width, height, { quality:this.quality / 100 });
		else
			writer.begin(0, 0, width, height);
		let from = 0;
		for (let y = 0; y < height; y += 4) {
			let to = from + (8 * width);
			writer.send(buffer.slice(from, to));
			from = to;
		}
		writer.end();
		return writer.bitmap.buffer;
	}
	copy0(block, pixels, width, height) {
		var buffer = new Uint16Array(block.buffer);	
		let v = 0;
		while (v < block.height) {
			let u = 0;
			while (u < block.width) {
				let x = block.x + u;
				let y = block.y + v;
				let src = ((v * block.width) + u);
				let dst = ((y * width) + x);
				pixels[dst] = buffer[src];
				u++;
			}	
			v++;
		}
	}
	copy90(block, pixels, width, height) {
		var buffer = new Uint16Array(block.buffer);	
		let v = 0;
		while (v < block.height) {
			let u = 0;
			while (u < block.width) {
				let x = height - 1 - block.y - v;
				let y = block.x + u;
				let src = ((v * block.width) + u);
				let dst = ((y * height) + x);
				pixels[dst] = buffer[src];
				u++;
			}	
			v++;
		}
	}
	copy180(block, pixels, width, height) {
		var buffer = new Uint16Array(block.buffer);	
		let v = 0;
		while (v < block.height) {
			let u = 0;
			while (u < block.width) {
				let x = width - 1 - block.x - u;
				let y = height - 1 - block.y - v;
				let src = ((v * block.width) + u);
				let dst = ((y * width) + x);
				pixels[dst] = buffer[src];
				u++;
			}	
			v++;
		}
	}
	copy270(block, pixels, width, height) {
		var buffer = new Uint16Array(block.buffer);	
		let v = 0;
		while (v < block.height) {
			let u = 0;
			while (u < block.width) {
				let x = block.y + v;
				let y = width - 1 - block.x - u;
				let src = ((v * block.width) + u);
				let dst = ((y * height) + x);
				pixels[dst] = buffer[src];
				u++;
			}	
			v++;
		}
	}
	pad(dimension) {
		let multiple = 4;
		let overflow = dimension & (multiple - 1);
		if (overflow)
			dimension += multiple - overflow;
		return dimension;
	}
	readGIF(buffer) @ "Tool_readGIF";
	readJPEG(path, frames) {
		var data = this.readFileBuffer(path);
		var jpeg = new JPEG(data, { pixelFormat:Bitmap.RGB565LE });
		var jpegWidth = jpeg.width;
		var jpegHeight = jpeg.height;
		if (this.quality == 0) {
			this.format = Bitmap.JPEG;
			frames.push(data);
			frames.width = jpegWidth;
			frames.height = jpegHeight;
			return;
		}
		var padWidth = this.pad(jpegWidth);
		var padHeight = this.pad(jpegHeight);
		var copy, width, height;
		switch (this.rotation) {
		case 0:
			copy = this.copy0;
			width = padWidth;
			height = padHeight;
			break;
		case 90:
			copy = this.copy90;
			width = padHeight;
			height = padWidth;
			break;
		case 180:
			copy = this.copy180;
			width = padWidth;
			height = padHeight;
			break;
		case 270:
			copy = this.copy270;
			width = padHeight;
			height = padWidth;
			break;
		}
		var pixels = new Uint16Array(width * height);
		for (;;) {
			var block = jpeg.read();
			if (!block)
				break;
			copy.call(this, block, pixels, padWidth, padHeight);
		}
		frames.push(this.compress(pixels.buffer, width, height));
		frames.width = width;
		frames.height = height;
	}
	readPNG(path, frames) {
		var data = this.readFileBuffer(path);
		var png = new PNG(data);
		let pngChannels = png.channels;
		let pngDepth = png.depth;
		let pngWidth = png.width;
		let pngHeight = png.height;
		if (((pngChannels != 3) && (pngChannels != 4)) || (pngDepth != 8))
			throw new Error("'" + pngPath + "': invalid PNG format!");
		if (this.quality == 0) {
			this.format = Bitmap.PNG;
			frames.push(data);
			frames.width = pngWidth;
			frames.height = pngHeight;
			return;
		}
		let padWidth = this.pad(pngWidth);
		let padHeight = this.pad(pngHeight);
		let bufferRGBA32 = new Uint8Array(padWidth * padHeight * 4);
		let offset = 0;
		let y = 0;
		while (y < pngHeight) {
			let pngLine = png.read();
			let pngOffset = 0;
			let x = 0;
			while (x < pngWidth) {
				bufferRGBA32[offset++] = pngLine[pngOffset++];
				bufferRGBA32[offset++] = pngLine[pngOffset++];
				bufferRGBA32[offset++] = pngLine[pngOffset++];
				bufferRGBA32[offset++] = (pngChannels == 4) ? pngLine[pngOffset++] : 255;
				x++;
			}
			while (x < padWidth) {
				bufferRGBA32[offset++] = 0;
				bufferRGBA32[offset++] = 0;
				bufferRGBA32[offset++] = 0;
				bufferRGBA32[offset++] = 255;
				x++;
			}
			y++;
		}
		while (y < padHeight) {
			let x = 0;
			while (x < padWidth) {
				bufferRGBA32[offset++] = 0;
				bufferRGBA32[offset++] = 0;
				bufferRGBA32[offset++] = 0;
				bufferRGBA32[offset++] = 255;
				x++;
			}
			y++;
		}
		let bufferRGB565LE = new Uint16Array(padWidth * padHeight);
		var colorConvert = new Convert(Bitmap.RGBA32, Bitmap.RGB565LE);
		colorConvert.process(bufferRGBA32.buffer, bufferRGB565LE.buffer);
		var pixels, width, height;
		switch (this.rotation) {
		case 0:
			pixels = bufferRGB565LE.buffer;
			width = padWidth;
			height = padHeight;
			break;
		case 90:
			pixels = this.rotate90(bufferRGB565LE.buffer, padWidth, padHeight);
			width = padHeight;
			height = padWidth;
			break;
		case 180:
			pixels = this.rotate180(bufferRGB565LE.buffer, padWidth, padHeight);
			width = padWidth;
			height = padHeight;
			break;
		case 270:
			pixels = this.rotate270(bufferRGB565LE.buffer, padWidth, padHeight);
			width = padHeight;
			height = padWidth;
			break;
		}
		frames.push(this.compress(pixels, width, height));
		frames.width = width;
		frames.height = height;
	}
	rotate90(pixels, width, height) {
		let buffer = new Uint16Array(pixels);
		let result = new Uint16Array(height * width);
		let v = 0;
		while (v < height) {
			let u = 0;
			while (u < width) {
				let x = height - 1 - v;
				let y = u;
				let src = ((v * width) + u);
				let dst = ((y * height) + x);
				result[dst] = buffer[src];
				u++;
			}	
			v++;
		}
		return result.buffer;
	}
	rotate180(pixels, width, height) {
		let buffer = new Uint16Array(pixels);
		let result = new Uint16Array(height * width);
		let v = 0;
		while (v < height) {
			let u = 0;
			while (u < width) {
				let x = width - 1 - u;
				let y = height - 1 - v;
				let src = ((v * width) + u);
				let dst = ((y * width) + x);
				result[dst] = buffer[src];
				u++;
			}	
			v++;
		}
		return result.buffer;
	}
	rotate270(pixels, width, height) {
		let buffer = new Uint16Array(pixels);
		let result = new Uint16Array(height * width);
		let v = 0;
		while (v < height) {
			let u = 0;
			while (u < width) {
				let x = v;
				let y = width - 1 - u;
				let src = ((v * width) + u);
				let dst = ((y * height) + x);
				result[dst] = buffer[src];
				u++;
			}	
			v++;
		}
		return result.buffer;
	}
	run() {
		var fps_numerator;
		var fps_denominator;
		var frames = new Array;
		if (this.directory) {
			var names = this.enumerateDirectory(this.inputPath);
			names.sort();
			var c = names.length;
			for (var i = 0; i < c; i++) {
				var path = this.inputPath + this.slash + names[i];
				var parts = this.splitPath(path);
				if (parts.extension == ".png") {
					this.readPNG(path, frames);
				}
				else if ((parts.extension == ".jpg") || (parts.extension == ".jpeg")) {
					this.readJPEG(path, frames);
				}
			}
			var parts = this.splitPath(this.inputPath);
			var a = parts.extension.match(/\.([0-9]+)fps/);
			fps_numerator = 1;
			fps_denominator = (a && (a.length == 2) && (a[1] !== undefined)) ? parseInt(a[1]) : 30;
		}
		else {
			var parts = this.splitPath(this.inputPath);
			if (parts.extension == ".gif") {
				var array = this.readGIF(this.readFileBuffer(this.inputPath));
				var gifWidth = array.width;
				var gifHeight = array.height;
				var gifDuration = array.reduce((sum, item) => sum + item.delay, 0);
				var width, height;
				switch (this.rotation) {
				case 0:
				case 180:
					width = gifWidth;
					height = gifHeight;
					break;
				case 90:
				case 270:
					width = gifHeight;
					height = gifWidth;
					break;
				}
				var c = array.length;
				for (var i = 0; i < c; i++) {
					var pixels;
					switch (this.rotation) {
					case 0:
						pixels = array[i];
						break;
					case 90:
						pixels = this.rotate90(array[i], gifWidth, gifHeight);
						break;
					case 180:
						pixels = this.rotate180(array[i], gifWidth, gifHeight);
						break;
					case 270:
						pixels = this.rotate270(array[i], gifWidth, gifHeight);
						break;
					}
					frames.push(this.compress(pixels, width, height));
				}
				frames.width = width;
				frames.height = height;
				fps_numerator = gifDuration / c;
				fps_denominator = 1000;
			}
			else if (parts.extension == ".png") {
				this.readPNG(this.inputPath, frames);
				fps_numerator = 1;
				fps_denominator = 1;
			}
			else if ((parts.extension == ".jpg") || (parts.extension == ".jpeg")) {
				this.readJPEG(this.inputPath, frames);
				fps_numerator = 1;
				fps_denominator = 1;
			}
			else
				throw new Error("'" + this.inputPath + "': invalid file!");
		}
		
		var frameCount = frames.length;
		var width = frames.width;
		var height = frames.height;
		parts.directory = this.outputPath;
		parts.extension = ".cs";
		if (this.quality !== undefined)
			parts.extension += this.quality;
		let output = new FILE(this.joinPath(parts), "wb");
		output.writeByte('c'.charCodeAt(0));
		output.writeByte('s'.charCodeAt(0));
		output.writeByte(this.format);
		output.writeByte(0);
		output.writeByte(width & 0xFF);
		output.writeByte(width >> 8);
		output.writeByte(height & 0xFF);
		output.writeByte(height >> 8);
		output.writeByte(frameCount & 0xFF);
		output.writeByte(frameCount >> 8);
		output.writeByte(fps_numerator & 0xFF);
		output.writeByte(fps_numerator >> 8);
		output.writeByte(fps_denominator & 0xFF);
		output.writeByte(fps_denominator >> 8);
		for (var frameIndex = 0; frameIndex < frameCount; frameIndex++) {
			var buffer = frames[frameIndex];
			var frameSize = buffer.byteLength;
			output.writeByte(frameSize & 255);
			output.writeByte(frameSize >> 8);
			output.writeBuffer(buffer);
		}
		// 0 size frame at the end
		output.writeByte(0);
		output.writeByte(0);
		output.close();
	}
}

File: tools/image2cs.c
/*
 * Copyright (c) 2016-2017  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK Tools.
 * 
 *   The Moddable SDK Tools is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 * 
 *   The Moddable SDK Tools is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 * 
 *   You should have received a copy of the GNU General Public License
 *   along with the Moddable SDK Tools.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "mc.xs.h"

typedef struct {
	uint16_t* begin;
	uint16_t* current;
	uint16_t* end;
	uint32_t delta;
	uint32_t length;
	uint16_t* from;
	uint16_t* to;
	uint16_t* first;
	uint16_t* last;
	int pass;
	int transparentIndex;
} txRow;

static int* bufferColorIndexes(int* prefixes, int* colorIndexes, int* buffer, int code, int clearCode)
{
	while (code > clearCode) {
		*buffer++ = colorIndexes[code];
		code = prefixes[code];
	}
	*buffer = code;
	return buffer;
}

static uint16_t pad(uint16_t dimension, uint16_t multiple)
{
	uint16_t overflow = dimension & (multiple - 1);
	if (overflow)
		dimension += multiple - overflow;
	return dimension;
}

static void poke(txRow* row, uint16_t* colorTable, int index)
{
    if ((row->from <= row->current) && (row->current < row->to)) {
        if (index == row->transparentIndex) {
            row->current++;
        }
        else {
            *row->current++ = colorTable[index];
        }
        if (row->current == row->end) {
          	if (row->pass == 0) {
            	row->begin += row->length;
 			}
         	else if (row->pass == 1) {
				row->begin += 8 * row->length;
				if (row->begin >= row->last) {
					row->begin = row->first + 4 * row->length;
					row->pass++;
				}
			}
			else if (row->pass == 2) {
				row->begin += 8 * row->length;
				if (row->begin >= row->last) {
					row->begin = row->first + 2 * row->length;
					row->pass++;
				}
			}
			else if (row->pass == 3) {
				row->begin += 4 * row->length;
				if (row->begin >= row->last) {
					row->begin = row->first + 1 * row->length;
					row->pass++;
				}
			}
			else if (row->pass == 4) {
				row->begin += 2 * row->length;
			}
            row->current = row->begin;
            row->end = row->begin + row->delta;
       }
    }
}

#define MAX_CODE 4096

#define BYTE ((bytes < bytesLimit) ? *bytes++ : (xsUnknownError("invalid GIF buffer"), 0))

void Tool_readGIF(xsMachine* the)
{
	static uint8_t byteMasks[8] = { 
		0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80 
	};
	static int codeMasks[16] = { 
		0x0001, 0x0002, 0x0004, 0x0008, 0x0010, 0x0020, 0x0040, 0x0080,
		0x0100, 0x0200, 0x0400, 0x0800, 0x1000, 0x2000, 0x4000, 0x8000
	};
    uint16_t globalColorTable[256];
    uint16_t localColorTable[256];

	uint8_t* bytesInitial;
	uint8_t* bytesLimit;
	uint8_t* bytes;
	uint32_t bufferSize;
	uint32_t bufferOffset;
	
	txRow row;
	
	uint8_t byte;
	uint16_t globalWidth;
	uint16_t globalHeight;
//	uint8_t backgroundColorIndex;
	
	uint8_t count;
	
	uint16_t delay = 0;
//	uint8_t disposalMethod = 0;
	uint8_t transparentColorFlag = 0;
	uint8_t transparentColorIndex = 0;
	
	uint16_t* screen;
	uint16_t* colorTable;
	
	xsVars(2);
	bufferSize = xsToInteger(xsGet(xsArg(0), xsID_byteLength));
	if (xsIsInstanceOf(xsArg(0), xsArrayBufferPrototype))
		bytesInitial = xsToArrayBuffer(xsArg(0));
	else
		bytesInitial = xsGetHostData(xsArg(0));
	bytesLimit = bytesInitial + bufferSize;
	bytes = bytesInitial;
	
	if (bufferSize < 6)
		xsUnknownError("invalid GIF buffer");
	if ((bytes[0] != 'G') || (bytes[1] != 'I') ||(bytes[2] != 'F'))
		xsUnknownError("invalid GIF signature");
	if ((bytes[3] != '8') || ((bytes[4] != '7') && (bytes[4] != '9')) || (bytes[5] != 'a'))
		xsUnknownError("invalid GIF version");
	bytes += 6;
	globalWidth = BYTE;
	globalWidth += BYTE << 8;
	globalHeight = BYTE;
	globalHeight += BYTE << 8;

	globalWidth = pad(globalWidth, 4);
	globalHeight = pad(globalHeight, 4);
	
	byte = BYTE;
	/* backgroundColorIndex = */ BYTE;
	bytes++;
	if (byte & 0x80) {
		int c = 2 << (byte & 0x07), i;
        for (i = 0; i < c; i++) {
        	uint16_t r = BYTE;
        	uint16_t g = BYTE;
        	uint16_t b = BYTE;
        	globalColorTable[i] = (((r >> 3) << 11) | ((g >> 2) << 5) | (b >> 3));
        }
	}
		
	bufferOffset = bytes - bytesInitial;
	xsResult = xsNewArray(0);
	xsDefine(xsResult, xsID_width, xsInteger((xsIntegerValue)globalWidth), xsDefault);
	xsDefine(xsResult, xsID_height, xsInteger((xsIntegerValue)globalHeight), xsDefault);
	if (xsIsInstanceOf(xsArg(0), xsArrayBufferPrototype))
		bytesInitial = xsToArrayBuffer(xsArg(0));
	else
		bytesInitial = xsGetHostData(xsArg(0));
	bytesLimit = bytesInitial + bufferSize;
	bytes = bytesInitial + bufferOffset;
		
	for (;;) {
		byte = BYTE;
		if (byte == 0x21) {
			byte = BYTE;
			if (byte == 0xF9) {
				bytes++;
				byte = BYTE;
//				disposalMethod = ((byte & 0x1C) >> 2);
				transparentColorFlag = byte & 0x01;
				delay = BYTE;
				delay += BYTE << 8;
				transparentColorIndex = BYTE;
				bytes++;
			}
			else if ((byte == 0x01) || (byte == 0xFE) || (byte == 0xFF)) {
				while ((count = BYTE))
					bytes += count;
			}
			else {
				xsUnknownError("invalid GIF extension");
			}
			
		}
		else if (byte== 0x2C) {
			int prefixes[MAX_CODE];
			int colorIndexes[MAX_CODE];
			int buffer[MAX_CODE];
			int* first = buffer;
			int* last;

			uint16_t x, y, width, height;
			uint8_t interlaceFlag;
	
			int colorIndexBitCount, codeBitCount, codeBitIndex, byteBitIndex;
			int codeAvailable, clearCode, endCode, nextCode, newCode, oldCode, colorIndex;
		
			bufferOffset = bytes - bytesInitial;
			xsVar(0) = xsArrayBuffer(NULL, (globalWidth * globalHeight * 2));
			xsDefine(xsVar(0), xsID_delay, xsInteger((xsIntegerValue)delay * 10), xsDefault);
			
			xsCall1(xsResult, xsID_push, xsVar(0));
			if (xsIsInstanceOf(xsArg(0), xsArrayBufferPrototype))
				bytesInitial = xsToArrayBuffer(xsArg(0));
			else
				bytesInitial = xsGetHostData(xsArg(0));
			bytesLimit = bytesInitial + bufferSize;
			bytes = bytesInitial + bufferOffset;
		
			screen = xsToArrayBuffer(xsVar(0));
			if (xsTest(xsVar(1))) {
				uint16_t* former = xsToArrayBuffer(xsVar(1));
				c_memcpy(screen, former, (globalWidth * globalHeight * 2));
			}
			else
				c_memset(screen, 0xFF, (globalWidth * globalHeight * 2));
			xsVar(1) = xsVar(0);
		
			x = BYTE;
			x += BYTE << 8;
			y = BYTE;
			y += BYTE << 8;
			width = BYTE;
			width += BYTE << 8;
			height = BYTE;
			height += BYTE << 8;
			byte = BYTE;
			interlaceFlag = (byte & 0x40) >> 6;
			if (byte & 0x80) {
				int c = 2 << (byte & 0x07), i;
				for (i = 0; i < c; i++) {
					uint16_t r = BYTE;
					uint16_t g = BYTE;
					uint16_t b = BYTE;
					localColorTable[i] = (((r >> 3) << 11) | ((g >> 2) << 5) | (b >> 3));
				}
				colorTable = localColorTable;
			}
			else
				colorTable = globalColorTable;
			
			row.begin = screen + (((y * globalWidth) + x));
			row.current = row.begin;
			row.end = row.begin + width;
			row.delta = width;
			row.length = globalWidth;
			row.from = screen;
			row.to = screen + (globalWidth * globalHeight);
			row.transparentIndex = (transparentColorFlag) ? transparentColorIndex : -1;
			if (interlaceFlag) {
				row.pass = 1;
				row.first = row.begin;
				row.last = row.begin + (height * row.length);
			}
			else
				row.pass = 0;

			colorIndexBitCount = BYTE;
			clearCode = 1 << colorIndexBitCount;
			endCode = clearCode + 1;
			
            codeAvailable = 1 << colorIndexBitCount;
			codeBitCount = colorIndexBitCount + 1;
			nextCode = clearCode + 2;
			oldCode = -1;
			
			newCode = 0;
			codeBitIndex = 0;
			while ((count = BYTE)) {
				while (count) {
					byte = BYTE;
					count--;
					for (byteBitIndex = 0; byteBitIndex < 8; byteBitIndex++) {
						if (byte & byteMasks[byteBitIndex])
							newCode |= codeMasks[codeBitIndex];
						codeBitIndex++;
						if (codeBitIndex == codeBitCount) {
							if (newCode == endCode) {
								bytes += count;
								count = 0;
								break;
							}
							else if (newCode == clearCode) {
								codeAvailable = (1 << colorIndexBitCount) - 1;
								codeBitCount = colorIndexBitCount + 1;
                                nextCode = clearCode + 2;
								oldCode = -1;
							}
                            else {
                                codeAvailable--;
                                if (oldCode == -1) {
                                    colorIndex = oldCode = newCode;
                                    poke(&row, colorTable, colorIndex);
                                }
                                else {
                                    if (newCode >= nextCode) {
                                        *first = colorIndex;
                                        last = bufferColorIndexes(prefixes, colorIndexes, first + 1, oldCode, clearCode);
                                    }
                                    else {
                                        last = bufferColorIndexes(prefixes, colorIndexes, first, newCode, clearCode);
                                    }
                                    colorIndex = *last;
                                    while (last >= first)
                                        poke(&row, colorTable, *last--);
                                    if (nextCode < MAX_CODE) {
                                        prefixes[nextCode] = oldCode;
                                        colorIndexes[nextCode] = colorIndex;
                                        nextCode++;
                                    }
                                    oldCode = newCode;
                                }
                            }
							if ((codeAvailable == 0) && (codeBitCount < 12)) {
								codeAvailable = 1 << codeBitCount;
								codeBitCount++;
							}
							codeBitIndex = 0;
                            newCode = 0;
						}
					}
				}
			}
		}
		else if (byte == 0x3B) {
			break;
		}
		else
			xsUnknownError("invalid GIF extension");
	}
}

